### PR TITLE
ifconfig: fix regex matching interface name with hyphen

### PR DESCRIPTION
**Pull request by dheerajd-msys**
Signed-off-by: dheerajd-msys <dheeraj.dubey@msystechnologies.com>

- Updated regex to match the interface name with hyphen 

## Description

- Updated regex will match with the output of `ifconfig` having device name `-` after any word character
 e.g. `br-1234`
- Ensured `chefstyle`
- Added test case.
- After regex update, data which is accommodated in array has shifted from index 2 to 3. Incorporated this change in commit https://github.com/chef/chef/pull/8756/commits/8e295ba6e3397add1ad26d92d97d66b58659424e 

## Related Issue

Fixes https://getchef.zendesk.com/agent/tickets/22294
## Error
`RHEL7: ifconfig provider fails with "undefined method [] for nil:NilClas`

## Recipe 
```
[root@ip-172-31-26-90 chef]# cat ../recipe-ifconfig.rb
ifconfig("52.34.83.20") do
action :add
device "br-1234"
onboot "yes"
mask "255.255.255.255"
target "52.34.83.20"
end
```
## Output after fix
```
[root@ip-172-31-26-90 chef]# bundle exec chef-apply ../recipe-ifconfig.rb
Recipe: (chef-apply cookbook)::(chef-apply recipe)
  * ifconfig[52.34.83.20] action add
    - run ifconfig br-1234 52.34.83.20 netmask 255.255.255.255 to add ifconfig[52.34.83.20]
  Recipe: <Dynamically Defined Resource>
    * file[/etc/sysconfig/network-scripts/ifcfg-br-1234] action create
      - update content in file /etc/sysconfig/network-scripts/ifcfg-br-1234 from ef289f to c40ecd
      --- /etc/sysconfig/network-scripts/ifcfg-br-1234  2019-07-19 09:52:51.246737514 +0000
      +++ /etc/sysconfig/network-scripts/.chef-ifcfg-br-123420190719-23908-1btlw8j      2019-07-19 09:55:56.679533307 +0000
      @@ -1,11 +1,20 @@
      +
       DEVICE=br-1234
      -STP=no
      -TYPE=Bridge
      -BOOTPROTO=none
      -IPADDR=52.34.83.20
      -PREFIX=32
      -IPV6INIT=no
      -NAME=br1
       ONBOOT=yes
      -DELAY=0
      +
      +IPADDR=52.34.83.20
      +NETMASK=255.255.255.255
      +
      +
      +
      +
      +
      +
      +
      +
      +
      +
      +
      +
      +
      - restore selinux security context

[root@ip-172-31-26-90 chef]#
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)
- [ ] Chore (non-breaking change that does not add functionality or fix an issue)

## Checklist:

- [x] I have read the **CONTRIBUTING** document.
- [x] I have run the pre-merge tests locally and they pass.
- [ ] I have updated the documentation accordingly.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.
- [x] All commits have been signed-off for [the Developer Certificate of Origin](https://github.com/chef/chef/blob/master/CONTRIBUTING.md#developer-certification-of-origin-dco).

### CODE REVIEW

File: lib/chef/provider/ifconfig.rb [hunk 1]
```diff
@@ -111,7 +111,7 @@ def load_current_resource
           #       TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
           @status = shell_out("ifconfig")
           @status.stdout.each_line do |line|
-            addr_regex = /^(\w+):?(\d*):?\ .+$/
+            addr_regex = /^(\w+)(-?):?(\d*):?\ .+$/
```
Comment (tas50) [inline lib/chef/provider/ifconfig.rb]: Can you attack a permalink from https://rubular.com/ that shows the various interface names we're trying to match so it's more clear to those in the future what this regex does?

Comment (dheerajd-msys) [reply lib/chef/provider/ifconfig.rb]: @tas50 Here is the list of interfaces that we are trying to match. Also added output of `ifconfig` so it's more clear.

Output of `ifconfig` after creating `Bridge interface` i.e. **br-1234**
```
[root@ip-172-31-26-90 chef]# ifconfig
br-1234: flags=4099<UP,BROADCAST,MULTICAST>  mtu 1500
        inet 52.34.83.20  netmask 255.255.255.255  broadcast 52.34.83.20
        ether 02:6a:d0:5d:0d:bd  txqueuelen 1000  (Ethernet)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

eth0: flags=4163<UP,BROADCAST,RUNNING,MULTICAST>  mtu 9001
        inet 172.31.26.90  netmask 255.255.240.0  broadcast 172.31.31.255
        inet6 fe80::2e:4eff:fe78:6c90  prefixlen 64  scopeid 0x20<link>
        ether 02:2e:4e:78:6c:90  txqueuelen 1000  (Ethernet)
        RX packets 454657  bytes 604056536 (576.0 MiB)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 87340  bytes 13338865 (12.7 MiB)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0

lo: flags=73<UP,LOOPBACK,RUNNING>  mtu 65536
        inet 127.0.0.1  netmask 255.0.0.0
        inet6 ::1  prefixlen 128  scopeid 0x10<host>
        loop  txqueuelen 1000  (Local Loopback)
        RX packets 0  bytes 0 (0.0 B)
        RX errors 0  dropped 0  overruns 0  frame 0
        TX packets 0  bytes 0 (0.0 B)
        TX errors 0  dropped 0 overruns 0  carrier 0  collisions 0
```

**Before fix permalink** : https://rubular.com/r/ujRptRHoxft64A
**After fix permalink** :  https://rubular.com/r/l9C8YlVgHO4cHN

Comment (tas50) [reply lib/chef/provider/ifconfig.rb]: Wouldn't it be more appropriate to use a match like https://rubular.com/r/JrykUpfjRnYeQD that would actually match on the full interface name.  Also can you  put  the actual permalink directly in the code comments so future developers can see what we're trying to match.

Comment (dheerajd-msys) [reply lib/chef/provider/ifconfig.rb]: Thank you @tas50, Updated!

**Comment by tas50** [comment]
@dheerajd-msys The functional tests are failing with this change

**Comment by dheerajd-msys** [comment]
Yes @tas50, Working on it now. Thank you :)

**Comment by tas50** [comment]
@dheerajd-msys Can you take a look at the various tests that are failing. I think they're probably unrelated, but we should get this green before merging. I believe you can fix the kitchen tests by switching this line to be 14 instead of current https://github.com/chef/chef/blob/chef-14/kitchen-tests/kitchen.yml#L6

**Comment by dheerajd-msys** [comment]
@tas50 PR https://github.com/chef/chef/pull/9405 is updated for kitchen tests. Still there are unrelated failures.

**Pull request merged**